src/actions/mod.rs

```diff
@@ -211,6 +213,49 @@ fn find_word_at_pos(line: &str, pos: &Column) -> (Column, Column) {
     (span::Column::new_zero_indexed(start), span::Column::new_zero_indexed(end))
 }
 
+// TODO include workspace Cargo.tomls in watchers / relevant
+/// Client file-watching request / filtering logic
+/// We want to watch workspace 'Cargo.toml', root 'Cargo.lock' & the root 'target' dir
+pub struct FileWatch<'ctx> {
+    project_str: &'ctx str,
+    project_uri: String,
+}
+
+impl<'ctx> FileWatch<'ctx> {
+    pub fn new(ctx: &'ctx InitActionContext) -> Self {
```

Comment:
I think it would be simpler to make FileWatch a module and make the two interesting methods functions which take `ctx` as an arg.

Reply:
This design allows fast calls of `is_relevant`. Notifications can include hundreds of thousands of file change references, so it makes sense to avoid allocations in the loop.

Reply:
OK, good point! Could you leave a comment about that somewhere please?

Reply:
Sure, I've added a note to the `is_relevant` method.